File: interfaces/matlab/examples/ocp/wave_energy/wave_energy.m
clear;

% Maximize produced wave energy
% The buoy moves using the wave energy (represented by a sin in hw)

BEGIN_ACADO;                                % Always start with "BEGIN_ACADO". 
    
    acadoSet('problemname', 'wave_energy'); % Set your problemname. If you 
                                            % skip this, all files will
                                            % be named "myAcadoProblem"
    
    
    DifferentialState       h;              % Position of the buoy
    DifferentialState       v;              % Velocity of the buoy
    DifferentialState       w;              % Produced wave energy
    Control                 u;              % The control (free parameter), can be used to adjust a force at the generator
    TIME                    t;              % the TIME (can be defined only once)

    h_hw = 10;                              % water level
    A_hw = 1.0;                             % amplitude of the waves
    T_hw = 5.0;                             % duration of a wave
    h_b  = 3.0;                             % height of the buoy
    rho  = 1000;                            % density of water
    A    = 1.0;                             % bottom area of the buoy
    m    = 100;                             % mass of the buoy
    g    = 9.81;                            % gravitational constant
    
    %% Intermediate state
    hw = h_hw + A_hw*sin(2*pi*t/T_hw);      % Height of the wave
  
    
    %% Differential Equation
    f = acado.DifferentialEquation();       % Set the differential equation object
    
    f.add(dot(h) ==  v);                    % Write down your ODE. You can add 
    f.add(dot(v) ==  rho*A*(hw-h)/m-g-u);   % multiple ODE's with f.add() when you 
    f.add(dot(w) ==  u*v);                  % have multiple differential states. 

    
    %% Optimal Control Problem
    ocp = acado.OCP(0.0, 15.0, 50);        % Set up the Optimal Control Problem (OCP) Start at 0s, control in 50 intervals to 15s
                                            
    ocp.maximizeMayerTerm( w );             % Maximize the energy
    
    ocp.subjectTo( f );                     % Your OCP is always subject to your 
                                            % differential equation
                                            
    ocp.subjectTo( 'AT_START', h - (h_hw-A_hw) ==  0.0 );
    ocp.subjectTo( 'AT_START', v ==  0.0 );
    ocp.subjectTo( 'AT_START', w ==  0.0 );
  
    ocp.subjectTo( -h_b <= h-hw <= 0.0 );
    ocp.subjectTo( 0.0 <= u <= 100.0 );
    
    
    %% Optimization Algorithm
    algo = acado.OptimizationAlgorithm(ocp); % Set up the optimization algorithm
    
    algo.set('MAX_NUM_ITERATIONS', 100 );    % Set some parameters for the algorithm
    algo.set('HESSIAN_APPROXIMATION','EXACT_HESSIAN' );
    
    
END_ACADO;           % Always end with "END_ACADO".
                     % This will generate a file problemname_ACADO.m. 
                     % Run this file to get your results. You can
                     % run the file problemname_ACADO.m as many
                     % times as you want without having to compile again.



out = wave_energy_RUN();                    % Run the test. The name of the RUN file
                                            % is problemname_RUN, so in
                                            % this case wave_energy_RUN
                                            
draw;

File: examples/ocp/wave_energy.cpp
/*
 *    This file is part of ACADO Toolkit.
 *
 *    ACADO Toolkit -- A Toolkit for Automatic Control and Dynamic Optimization.
 *    Copyright (C) 2008-2014 by Boris Houska, Hans Joachim Ferreau,
 *    Milan Vukov, Rien Quirynen, KU Leuven.
 *    Developed within the Optimization in Engineering Center (OPTEC)
 *    under supervision of Moritz Diehl. All rights reserved.
 *
 *    ACADO Toolkit is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 3 of the License, or (at your option) any later version.
 *
 *    ACADO Toolkit is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with ACADO Toolkit; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */


 /**
 *    \file examples/integrator/harmonic_oscillator.cpp
 *    \author Boris Houska, Hans Joachim Ferreau
 *    \date 2008
 */


#include <acado_optimal_control.hpp>
#include <acado_gnuplot.hpp>


int main( ){

  USING_NAMESPACE_ACADO;

  // Parameters
  double h_hw = 10;    // water level
  double A_hw = 1.0;   // amplitude of the waves
  double T_hw = 5.0;   // duration of a wave
  double h_b  = 3.0;   // height of the buoy
  double rho  = 1000;  // density of water
  double A    = 1.0;   // bottom area of the buoy
  double m    = 100;   // mass of the buoy
  double g    = 9.81;  // gravitational constant

  // Free parameter
  Control u;

  // Variables
  DifferentialState h; // Position of the buoy
  DifferentialState v; // Velocity of the buoy
  DifferentialState w; // Produced wave energy
  TIME t;

  // Differential equation
  DifferentialEquation f;
  
  // Height of the wave
  IntermediateState hw;
  hw = h_hw + A_hw*sin(2*M_PI*t/T_hw);
  
  f << dot(h) ==  v;
  f << dot(v) ==  rho*A*(hw-h)/m - g - u;
  f << dot(w) ==  u*v;

  // DEFINE AN OPTIMAL CONTROL PROBLEM:
  // ----------------------------------

  const double t_start    =  0.0        ;
  const double t_end      =  15;

  OCP ocp( t_start, t_end, 100 );
  ocp.maximizeMayerTerm( w );
  ocp.subjectTo( f );
  
//   double x_start[3];
//   x_start[0] = h_hw - 0*A_hw;
//   x_start[1] = 0;
//   x_start[2] = 0;

  ocp.subjectTo( AT_START, h - (h_hw-A_hw) ==  0.0 );
  ocp.subjectTo( AT_START, v ==  0.0 );
  ocp.subjectTo( AT_START, w ==  0.0 );
  
  ocp.subjectTo( -h_b <= h-hw <= 0.0 );
  ocp.subjectTo( 0.0 <= u <= 100.0 );
  
  

      // DEFINE A PLOT WINDOW:
    // ---------------------
  GnuplotWindow window;
  window.addSubplot( h,"Height of buoy" );
  window.addSubplot( v,"Velocity of buoy" );
  window.addSubplot( w,"Objective function " );
  window.addSubplot( u,"Resistance" );
  window.addSubplot( hw,"Wave height" );
  //  window.addSubplot( PLOT_KKT_TOLERANCE,"KKT Tolerance" );

  // DEFINE AN OPTIMIZATION ALGORITHM AND SOLVE THE OCP:
  // ---------------------------------------------------
  OptimizationAlgorithm algorithm(ocp);
  algorithm.set( HESSIAN_APPROXIMATION, EXACT_HESSIAN );
  algorithm.set( MAX_NUM_ITERATIONS, 100 );
  //  algorithm.set( KKT_TOLERANCE, 1e-10 );
  
  algorithm << window;
  algorithm.solve();

  return 0;
}



